[AArch64][GlobalISel] Add support for neon right shifts

JoshdRod opened the pull request:
Many neon right shift intrinsics were not supported by GlobalISel, mainly due to a lack of legalisation logic. This logic has now been implemented.

JoshdRod committed:
[AArch64][GlobalISel] Removed sqshl fallback occurring for <1 x i64> operands

GISel now places sqshl operands on floating point registers. Generated code is slightly less efficient compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for uqsh, uqrsh, and sqrsh with <1 x i64> operands

GISel now places operands for these intrinsics on floating point registers. Generated code is slightly less efficient compared to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for sqshlu intrinsic

Added G_SQSHLU node, which lowers the llvm ir intrinsic aarch64_neon_sqshlu to the machine intrinsic sqshlu. Generated code is slightly less efficient compare to SDAG.

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback from sqshrn intrinsic

In legalisation, the IR intrinsic is lowered to two GI instructions: a vector right shift (G_VASHR), and a truncate with saturation (G_TRUNC_SSAT_S). The result of the G_VASHR is the operand of the G_TRUNC_SSAT_S. Vectors that are treated as i64/i32 are dealt with in TableGen, so are not handled here.

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback from sqshrun intrinsic

In legalisation, the IR intrinsic is lowered to two GI instructions: a vector right shift (G_VASHR), and an unsigned truncate with saturation (G_TRUNC_SSAT_U). The result of the G_VASHR is the operand of the G_TRUNC_SSAT_U.

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for sqrshrn intrinsic

GISel now legalises sqrshrn into G_TRUNC_SSAT_S(G_SRSHR(vec, shift)).

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for sqrshrun intrinsic

GlobalISel now legalises sqrshrun to G_TRUNC_SSAT_U(G_SRSHR(vec, shift)).

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for uqrshrn intrinsic

GlobalISel now lowers uqrshrn to G_TRUNC_USATU(G_URSHR(vec, shift)).

JoshdRod committed:
[AArch64][GlobalISel] Removed fallback for uqshrn intrinsic

GlobalISel now lowers uqshrn to G_TRUNC_USATU(VLSHR(vec, shift)).

JoshdRod committed:
[AArch64][GlobalISel] Updated test checks

JoshdRod committed:
[AArch64][GlobalISel] Updated test checks

JoshdRod committed:
[AArch64][GlobalISel] Cleaned up test checks

JoshdRod committed:
[AArch64][GlobalISel] Fixed formatting

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1857,6 +1857,98 @@ bool AArch64LegalizerInfo::legalizeIntrinsic(LegalizerHelper &Helper,
+    break;
+  }
+  case Intrinsic::aarch64_neon_sqshrun: {
+    if (MRI.getType(MI.getOperand(0).getReg()).isVector()) {
```
Early return and reduce indent?

JoshdRod replied:
Resolved, let me know if there are any issues :+1:

JoshdRod committed:
[AArch64][GlobalISel] Refactored Legaliser Changes

Early returns implemented where possible, in order to reduce
indentation.

davemgreen commented on llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:
```diff
@@ -483,6 +483,20 @@ static bool isFPIntrinsic(const MachineRegisterInfo &MRI,
   case Intrinsic::aarch64_neon_sqadd:
   case Intrinsic::aarch64_neon_sqsub:
   case Intrinsic::aarch64_crypto_sha1h:
```
Keep the crypto next to each other.

JoshdRod committed:
[AArch64][GlobalISel] Formatting changes

JoshdRod committed:
[AArch64][GlobalISel] Renamed GI nodes describing intrinsics with immediate offsets

In SDAG, nodes that expect an immediate offset end with _I. This is now reflected in GISel for the vector shifts.

davemgreen reviewed (approved):
Can you rebase to remove the SLI/SRI changes from here? Otherwise I think this LGTM. Thanks.

JoshdRod committed:
[AArch64][GlobalISel] Modified comments to improve readability

cofibrant commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1857,6 +1857,111 @@ bool AArch64LegalizerInfo::legalizeIntrinsic(LegalizerHelper &Helper,
+    if (!MRI.getType(MI.getOperand(0).getReg()).isVector())
+      return false;
+    // Create right shift instruction. Get v. register the output is written
+    // to
```
I maybe would have said 'applied to' instead of 'taking in' but otherwise looks good

JoshdRod committed:
[AArch64][GlobalISel] Fix formatting

Some indentation errors were introduced in previous commits. This patch ran clang-format to find and fix them.
